compiler/src/dotty/tools/dotc/macros/Typer.scala

```diff
@@ -0,0 +1,95 @@
+package dotty.tools.dotc
+package macros
+
+import core._
+import ast._
+import Trees._
+import Contexts.Context
+import Types._
+import StdNames._
+import Decorators._
+import Constants._
+import Symbols._
+import Flags._
+
+/** Typer for typing meta blocks
+ *
+ *    http://docs.scala-lang.org/sips/pending/inline-meta.html
+ *
+ *  Inside a meta block, e.g.
+ *
+ *    inline def is[T](a: Any): Any = meta {
+ *      q"$a.isInstanceOf[$T]"
+ *    }
+ *
+ *  The variables will get following types:
+ *
+ *    T : scala.meta.Type
+ *    a : scala.meta.Term
+ *
+ *  If it's an annotation macro, e.g.
+ *
+ *    class main extends StaticAnnotation {
+ *      inline def apply(defn: Any): Any = meta { ... }
+ *    }
+ *
+ *  The parameter `defn` gets the type `scala.meta.Tree` inside the meta block.
```

Comment:
I think that the current implementation in scalameta/paradise uses `scala.meta.Stat`.